### PR TITLE
MON-4034: Add TelemeterClientConfig to ClusterMonitoring API

### PR DESCRIPTION
**Pull request by danielmellado**
Migrate the telemeter-client configmap settings to a CRD field within
ClusterMonitoringSpec in config/v1alpha1. The new TelemeterClientConfig
struct supports:

- nodeSelector: pod scheduling to specific nodes
- resources: compute resource requests and limits
- tolerations: pod tolerations for scheduling
- topologySpreadConstraints: pod distribution across topology domains

Signed-off-by: Daniel Mellado <dmellado@fedoraproject.org>

### CODE REVIEW

**Review by saschagrunert — CHANGES_REQUESTED** [review]
Creating a new review, tools messed this one up.

File: config/v1alpha1/types_cluster_monitoring.go [hunk 1]
```diff
@@ -107,6 +107,12 @@ type ClusterMonitoringSpec struct {
 	// When omitted, this means no opinion and the platform is left to choose a reasonable default, which is subject to change over time.
 	// +optional
 	PrometheusOperatorAdmissionWebhookConfig PrometheusOperatorAdmissionWebhookConfig `json:"prometheusOperatorAdmissionWebhookConfig,omitempty,omitzero"`
+	// telemeterClientConfig is an optional field that can be used to configure the Telemeter Client
+	// component that runs in the openshift-monitoring namespace. The Telemeter Client collects
+	// selected monitoring metrics and forwards them to Red Hat for telemetry purposes.
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default, which is subject to change over time.
+	// +optional
+	TelemeterClientConfig TelemeterClientConfig `json:"telemeterClientConfig,omitempty,omitzero"`
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:135]: This replaces `openShiftStateMetricsConfig` with `telemeterClientConfig`. On master, `OpenShiftStateMetricsConfig` was merged one day before this PR (4cf5fe17e, MON-4033). After rebasing, both fields need to coexist.

Likely happened because the PR was branched before MON-4033 merged. After rebase the result should be:

```go
	OpenShiftStateMetricsConfig OpenShiftStateMetricsConfig `json:"openShiftStateMetricsConfig,omitempty,omitzero"`
	TelemeterClientConfig TelemeterClientConfig `json:"telemeterClientConfig,omitempty,omitzero"`
```

File: config/v1alpha1/types_cluster_monitoring.go [hunk 2]
```diff
@@ -566,6 +572,80 @@ type PrometheusOperatorAdmissionWebhookConfig struct {
 	TopologySpreadConstraints []v1.TopologySpreadConstraint `json:"topologySpreadConstraints,omitempty"`
 }
 
+// TelemeterClientConfig provides configuration options for the Telemeter Client component
+// that runs in the `openshift-monitoring` namespace. The Telemeter Client collects selected
+// monitoring metrics and forwards them to Red Hat for telemetry purposes.
+// Use this configuration to control pod scheduling and resource allocation.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go]: Nit: "Use this configuration to control pod scheduling and resource allocation." is not present on any sibling types (`PrometheusOperatorConfig`, `MetricsServerConfig`, `OpenShiftStateMetricsConfig`). Consider removing for consistency.

File: config/v1alpha1/types_cluster_monitoring.go [hunk 3]
```diff
@@ -566,6 +572,80 @@ type PrometheusOperatorAdmissionWebhookConfig struct {
...
+// +kubebuilder:validation:MinProperties=1
+type TelemeterClientConfig struct {
+	// nodeSelector defines the nodes on which the Pods are scheduled.
+	// nodeSelector is optional.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:1764]: Nit: "nodeSelector is optional." repeats the field name. The `resources` field in this same struct uses "This field is optional." instead, which reads better. Same applies to `tolerations is optional.` at line 616.

File: config/v1alpha1/types_cluster_monitoring.go [hunk 4]
```diff
@@ -566,6 +572,80 @@ type PrometheusOperatorAdmissionWebhookConfig struct {
...
+	// This field is optional.
+	// More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+	// This is a simplified API that maps to Kubernetes ResourceRequirements.
+	// The current default values are:
+	//   resources:
+	//    - name: cpu
+	//      request: 1m
+	//      limit: null
+	//    - name: memory
+	//      request: 40Mi
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:1786]: The `OpenShiftStateMetricsConfig` this was copied from documents `32Mi` for default memory. Is `40Mi` the correct default for the Telemeter Client in the controller?

File: config/v1alpha1/types_cluster_monitoring.go [hunk 5]
```diff
@@ -566,6 +572,80 @@ type PrometheusOperatorAdmissionWebhookConfig struct {
...
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=10
+	// +kubebuilder:validation:MinItems=1
+	Resources []ContainerResource `json:"resources,omitempty"`
+	// tolerations defines tolerations for the pods.
+	// tolerations is optional.
+	//
+	// When omitted, this means the user has no opinion and the platform is left
+	// to choose reasonable defaults. These defaults are subject to change over time.
+	// Defaults are empty/unset.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:1802]: Nit: "These defaults are subject to change over time. Defaults are empty/unset." reads awkwardly. Consider: "The current default is an empty list."

File: config/v1alpha1/tests/clustermonitorings.config.openshift.io/ClusterMonitoringConfig.yaml [hunk 6]
```diff
@@ -577,3 +577,256 @@ tests:
...
+            topologySpreadConstraints: []
+      expectedError: 'spec.telemeterClientConfig.topologySpreadConstraints: Invalid value: 0: spec.telemeterClientConfig.topologySpreadConstraints in body should have at least 1 items'
+    - name: Should reject TelemeterClientConfig with empty resources array
+      initial: |
+        apiVersion: config.openshift.io/v1alpha1
+        kind: ClusterMonitoring
+        spec:
+          telemeterClientConfig:
+            resources: []
+      expectedError: 'spec.telemeterClientConfig.resources: Invalid value: 0: spec.telemeterClientConfig.resources in body should have at least 1 items'
```
Comment (saschagrunert) [inline config/v1alpha1/tests/clustermonitorings.config.openshift.io/ClusterMonitoringConfig.yaml:1020]: Missing boundary tests for `nodeSelector` and `tolerations`. Please add:
- Empty `nodeSelector: {}` (should fail MinProperties=1)
- nodeSelector with >10 entries (should fail MaxProperties=10)
- Empty `tolerations: []` (should fail MinItems=1)
- tolerations with >10 entries (should fail MaxItems=10)

**Review by saschagrunert — CHANGES_REQUESTED** [review]
The new `TelemeterClientConfig` struct follows the established pattern from other component configs in this file. Markers and godoc are mostly well done.

Main issue: the PR removes `OpenShiftStateMetricsConfig` (merged to master one day prior via MON-4033, 4cf5fe17e) and replaces it with `TelemeterClientConfig`. After rebasing on current master, both fields need to coexist.

Also needs additional test coverage for `nodeSelector` and `tolerations` boundary validation.

File: config/v1alpha1/types_cluster_monitoring.go (hunk 1 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go:135]: Rebased on current master. Both openShiftStateMetricsConfig and telemeterClientConfig now coexist in ClusterMonitoringSpec.

File: config/v1alpha1/types_cluster_monitoring.go (hunk 2 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go]: Good catch, removed for consistency with the sibling types.

File: config/v1alpha1/types_cluster_monitoring.go (hunk 3 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go:1764]: Agreed, changed both to "This field is optional." to match the resources field in the same struct.

File: config/v1alpha1/types_cluster_monitoring.go (hunk 4 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go:1786]: Yes, 40Mi is correct for the Telemeter Client. Confirmed from the controller's deployment manifest (cluster-monitoring-operator/assets/telemeter-client/deployment.yaml). The 32Mi in OpenShiftStateMetricsConfig is specific to that component — each component has its own defaults.

File: config/v1alpha1/types_cluster_monitoring.go (hunk 5 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go:1802]: Fixed, now reads "The current default is an empty list."

File: config/v1alpha1/tests/clustermonitorings.config.openshift.io/ClusterMonitoringConfig.yaml (hunk 6 above)
Comment (danielmellado) [reply config/v1alpha1/tests/clustermonitorings.config.openshift.io/ClusterMonitoringConfig.yaml:1020]: Added all four boundary tests:

Empty nodeSelector: {} — rejected by MinProperties=1
nodeSelector with 11 entries — rejected by MaxProperties=10
Empty tolerations: [] — rejected by MinItems=1
tolerations with 11 entries — rejected by MaxItems=10

**Comment by danielmellado** [comment]
While rebasing, the CRD hit the Kubernetes CEL validation cost budget (StaticEstimatedCRDCostLimit). The quantity() CEL function has a high fixed estimated cost per invocation, and with 6 structs in ClusterMonitoringSpec each embedding []ContainerResource, the cumulative cost exceeded the limit by ~13%.

Fixed by reducing MaxItems on all resources []ContainerResource fields from 10 to 5 (sufficient for practical use — cpu, memory, hugepages variants).I also added an explanatory comment on the ContainerResource type.

File: config/v1alpha1/types_cluster_monitoring.go [hunk 7]
```diff
@@ -309,6 +315,12 @@ const (
 )
 
 // ContainerResource defines a single resource requirement for a container.
+// MaxItems on []ContainerResource fields is kept at 5 to stay within the
+// Kubernetes CRD CEL validation cost budget (StaticEstimatedCRDCostLimit).
+// The quantity() CEL function has a high fixed estimated cost per invocation,
+// and the limit-vs-request comparison rule is costed per maxItems per location.
+// With multiple structs in ClusterMonitoringSpec embedding []ContainerResource,
+// maxItems > 5 causes the total estimated rule cost to exceed the budget.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:407]: This is an internal implementation detail that appears in the CRD description and would show up in oc explain. Per the OpenShift API conventions, godoc should be user-readable documentation. Users don't need to know about CEL cost budgets. This should be moved behind a `// + ---` separator so it doesn't appear in the generated schema, or removed entirely from the type comment and placed in a separate developer-only comment block.

File: config/v1alpha1/types_cluster_monitoring.go [hunk 8]
```diff
@@ -202,13 +208,13 @@ type AlertmanagerCustomConfig struct {
...
 	//      limit: null
-	// Maximum length for this list is 10.
+	// Maximum length for this list is 5.
 	// Minimum length for this list is 1.
 	// Each resource name must be unique within this list.
 	// +optional
 	// +listType=map
 	// +listMapKey=name
-	// +kubebuilder:validation:MaxItems=10
+	// +kubebuilder:validation:MaxItems=5
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:304]: Whenever we tighten a validation, it is imperative that we ensure that writes to existing objects continue to work. We do this through ratcheting. We expect a ratcheting validation test to be added whenever we tighten validation.

File: config/v1alpha1/types_cluster_monitoring.go [hunk 9]
```diff
@@ -645,6 +657,79 @@ type OpenShiftStateMetricsConfig struct {
...
+	// Maximum length for this list is 10.
+	// Minimum length for this list is 1.
+	// +kubebuilder:validation:MaxItems=10
+	// +kubebuilder:validation:MinItems=1
+	// +listType=atomic
+	// +optional
+	Tolerations []v1.Toleration `json:"tolerations,omitempty"`
+	// topologySpreadConstraints defines rules for how Telemeter Client Pods should be distributed
+	// across topology domains such as zones, nodes, or other user-defined labels.
+	// topologySpreadConstraints is optional.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:1812]: To achieve maximum consistency: 
```suggestion
	// This field is optional.
```

File: config/v1alpha1/types_cluster_monitoring.go [hunk 10]
```diff
@@ -645,6 +657,79 @@ type OpenShiftStateMetricsConfig struct {
...
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=5
+	// +kubebuilder:validation:MinItems=1
+	Resources []ContainerResource `json:"resources,omitempty"`
+	// tolerations defines tolerations for the pods.
+	// This field is optional.
+	//
+	// When omitted, this means the user has no opinion and the platform is left
+	// to choose reasonable defaults. These defaults are subject to change over time.
+	// The current default is an empty list.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go]: Another consistency nit:
```suggestion
	// Default is empty list.
```

File: config/v1alpha1/types_cluster_monitoring.go (hunk 7 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go:407]: Good point. Moved the implementation detail behind a // --- separator so it no longer appears in the generated CRD description or oc explain output. The comment is still visible to developers in the source.

File: config/v1alpha1/types_cluster_monitoring.go (hunk 8 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go:304]: Added three ratcheting onUpdate tests using initialCRDPatches (temporarily setting maxItems back to 10 to create an object with 7 resources under the old schema):

1. Can update other fields while retaining >5 resources (ratcheting preserves the old value)
2. Cannot change to a different invalid list (still >5 items, rejected)
3. Can reduce to a valid count (<=5 items, accepted)

Would that be enough?

File: config/v1alpha1/types_cluster_monitoring.go (hunk 9 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go:1812]: Fixed

File: config/v1alpha1/types_cluster_monitoring.go (hunk 10 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go]: Same as above, thanks!

**Comment by danielmellado** [comment]
The failure on `verify-crdify`is to be expected as per the CEL fix
```
		(v1alpha1) ^.spec.prometheusOperatorAdmissionWebhookConfig.resources - maxItems: 
              
			maximum decreased : 10 -> 5
+ echo This verifier checks all files that have changed. In some cases you may have changed or renamed a file that already contained api violations, but you are not introducing a new violation. In such cases it is appropriate to /override the failing CI job.
```

File: config/v1alpha1/types_cluster_monitoring.go [hunk 11]
```diff
@@ -645,6 +658,79 @@ type OpenShiftStateMetricsConfig struct {
 	TopologySpreadConstraints []v1.TopologySpreadConstraint `json:"topologySpreadConstraints,omitempty"`
 }
 
+// TelemeterClientConfig provides configuration options for the Telemeter Client component
+// that runs in the `openshift-monitoring` namespace. The Telemeter Client collects selected
+// monitoring metrics and forwards them to Red Hat for telemetry purposes.
+// +kubebuilder:validation:MinProperties=1
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:1761]: The MinProperties=1 constraint is not documented in the type comment. All validation markers must be reflected in the comment per API conventions.

```suggestion
// At least one field must be specified.
// +kubebuilder:validation:MinProperties=1
```

**Comment by danielmellado** [comment]
Thanks, I'm committing this, rebasing and I'll push the updated version ;)

File: config/v1alpha1/types_cluster_monitoring.go [hunk 12]
```diff
@@ -1734,6 +1753,80 @@ const (
...
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=5
+	// +kubebuilder:validation:MinItems=1
+	Resources []ContainerResource `json:"resources,omitempty"`
+	// tolerations defines tolerations for the pods.
+	// This field is optional.
+	//
+	// When omitted, this means the user has no opinion and the platform is left
+	// to choose reasonable defaults. These defaults are subject to change over time.
+	// Default is empty list.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go]: nit:

```suggestion
	// Defaults are empty/unset.
```

File: config/v1alpha1/types_cluster_monitoring.go [hunk 13]
```diff
@@ -1734,6 +1753,80 @@ const (
...
+	Tolerations []v1.Toleration `json:"tolerations,omitempty"`
+	// topologySpreadConstraints defines rules for how Telemeter Client Pods should be distributed
+	// across topology domains such as zones, nodes, or other user-defined labels.
+	// This field is optional.
+	// This helps improve high availability and resource efficiency by avoiding placing
+	// too many replicas in the same failure domain.
+	//
+	// When omitted, this means no opinion and the platform is left to choose a default, which is subject to change over time.
+	// This field maps directly to the `topologySpreadConstraints` field in the Pod spec.
+	// Default is empty list.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:1818]: nit:

```suggestion
	// Defaults are empty/unset.
```

File: config/v1alpha1/types_cluster_monitoring.go [hunk 14]
```diff
@@ -1734,6 +1753,80 @@ const (
...
 )
 
+// TelemeterClientConfig provides configuration options for the Telemeter Client component
+// that runs in the `openshift-monitoring` namespace. The Telemeter Client collects selected
+// monitoring metrics and forwards them to Red Hat for telemetry purposes.
+// At least one field must be specified.
+// +kubebuilder:validation:MinProperties=1
+type TelemeterClientConfig struct {
+	// nodeSelector defines the nodes on which the Pods are scheduled.
+	// This field is optional.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go]: nit:

```suggestion
	// nodeSelector is optional.
```

File: config/v1alpha1/types_cluster_monitoring.go [hunk 15]
```diff
@@ -1734,6 +1753,80 @@ const (
...
+	// Minimum length for this list is 1.
+	// Each resource name must be unique within this list.
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=5
+	// +kubebuilder:validation:MinItems=1
+	Resources []ContainerResource `json:"resources,omitempty"`
+	// tolerations defines tolerations for the pods.
+	// This field is optional.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go]: nit:

```suggestion
	// tolerations is optional.
```

File: config/v1alpha1/types_cluster_monitoring.go [hunk 16]
```diff
@@ -1734,6 +1753,80 @@ const (
...
+	// Maximum length for this list is 10.
+	// Minimum length for this list is 1.
+	// +kubebuilder:validation:MaxItems=10
+	// +kubebuilder:validation:MinItems=1
+	// +listType=atomic
+	// +optional
+	Tolerations []v1.Toleration `json:"tolerations,omitempty"`
+	// topologySpreadConstraints defines rules for how Telemeter Client Pods should be distributed
+	// across topology domains such as zones, nodes, or other user-defined labels.
+	// This field is optional.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go]:
```suggestion
	// topologySpreadConstraints is optional.
```

File: config/v1alpha1/types_cluster_monitoring.go [hunk 17]
```diff
@@ -1734,6 +1753,80 @@ const (
...
+	// The current default value is `kubernetes.io/os: linux`.
+	// When specified, nodeSelector must contain at least 1 entry and must not contain more than 10 entries.
+	// +optional
+	// +kubebuilder:validation:MinProperties=1
+	// +kubebuilder:validation:MaxProperties=10
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+	// resources defines the compute resource requests and limits for the Telemeter Client container.
+	// This includes CPU, memory and HugePages constraints to help control scheduling and resource usage.
+	// When not specified, defaults are used by the platform. Requests cannot exceed limits.
+	// This field is optional.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:1777]: This is fine since that matches all sibling structs.

File: config/v1alpha1/types_cluster_monitoring.go (hunk 12 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go]: I'll mark all these as resolved, thanks!

File: config/v1alpha1/types_cluster_monitoring.go [hunk 18]
```diff
@@ -288,13 +295,13 @@ type AlertmanagerCustomConfig struct {
...
 	//      limit: null
-	// Maximum length for this list is 10.
+	// Maximum length for this list is 5.
 	// Minimum length for this list is 1.
 	// Each resource name must be unique within this list.
 	// +optional
 	// +listType=map
 	// +listMapKey=name
-	// +kubebuilder:validation:MaxItems=10
+	// +kubebuilder:validation:MaxItems=5
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:304]: The ratcheting tests added only cover `alertmanagerConfig`. The `MaxItems` tightening from 10 to 5 also affects `metricsServerConfig`, `prometheusOperatorConfig`, `prometheusOperatorAdmissionWebhookConfig`, and `openShiftStateMetricsConfig`. Please add ratcheting tests with `initialCRDPatches` for those as well.

File: config/v1alpha1/types_cluster_monitoring.go [hunk 19]
```diff
@@ -393,6 +400,13 @@ const (
 )
 
 // ContainerResource defines a single resource requirement for a container.
+// ---
+// MaxItems on []ContainerResource fields is kept at 5 to stay within the
+// Kubernetes CRD CEL validation cost budget (StaticEstimatedCRDCostLimit).
+// The quantity() CEL function has a high fixed estimated cost per invocation,
+// and the limit-vs-request comparison rule is costed per maxItems per location.
+// With multiple structs in ClusterMonitoringSpec embedding []ContainerResource,
+// maxItems > 5 causes the total estimated rule cost to exceed the budget.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:407]: nit: The `---` separator does not prevent this from appearing in `oc explain`. Move the cost budget explanation to a non-godoc comment block (separated by a blank line above the godoc) so it does not leak into user-facing documentation.

Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go:407]: Done, thanks!

File: config/v1alpha1/types_cluster_monitoring.go (hunk 18 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go:304]: Added!

File: config/v1alpha1/types_cluster_monitoring.go [hunk 20]
```diff
@@ -762,17 +776,22 @@ type PrometheusConfig struct {
...
 	//      request: 40Mi
+	//      limit: null
+	// Maximum length for this list is 5.
+	// Minimum length for this list is 1.
+	// Each resource name must be unique within this list.
 	// +optional
 	// +listType=map
 	// +listMapKey=name
-	// +kubebuilder:validation:MaxItems=10
+	// +kubebuilder:validation:MaxItems=5
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go:795]: Missing ratcheting test for `PrometheusConfig.Resources`. All other component configs that had MaxItems tightened from 10 to 5 have ratcheting tests (alertmanagerConfig, metricsServerConfig, prometheusOperatorConfig, prometheusOperatorAdmissionWebhookConfig, openShiftStateMetricsConfig). This is the only one missing. Please add the same three-test pattern (retain via ratcheting, reject changed invalid list, allow reducing to valid count) with `initialCRDPatches` on `prometheusConfig/properties/resources/maxItems`.

File: config/v1alpha1/types_cluster_monitoring.go [hunk 21]
```diff
@@ -1734,6 +1753,80 @@ const (
...
+	Tolerations []v1.Toleration `json:"tolerations,omitempty"`
+	// topologySpreadConstraints defines rules for how Telemeter Client Pods should be distributed
+	// across topology domains such as zones, nodes, or other user-defined labels.
+	// topologySpreadConstraints is optional.
+	// This helps improve high availability and resource efficiency by avoiding placing
+	// too many replicas in the same failure domain.
+	//
+	// When omitted, this means no opinion and the platform is left to choose a default, which is subject to change over time.
+	// This field maps directly to the `topologySpreadConstraints` field in the Pod spec.
+	// Defaults are empty/unset.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go]: nit: All sibling structs use "Default is empty list." for `topologySpreadConstraints`. "Defaults are empty/unset." is the `tolerations` pattern.

```suggestion
	// Default is empty list.
```

**Review by saschagrunert — CHANGES_REQUESTED** [review]
The new TelemeterClientConfig struct follows established patterns well. Markers, godoc, and test coverage are mostly in good shape. Three items remain.

File: config/v1alpha1/types_cluster_monitoring.go [hunk 22]
```diff
@@ -762,17 +776,22 @@ type PrometheusConfig struct {
 	// This includes CPU, memory and HugePages constraints to help control scheduling and resource usage.
 	// When not specified, defaults are used by the platform. Requests cannot exceed limits.
 	// Each entry must have a unique resource name.
-	// Minimum of 1 and maximum of 10 resource entries can be specified.
+	// Minimum of 1 and maximum of 5 resource entries can be specified.
```
Comment (saschagrunert) [inline config/v1alpha1/types_cluster_monitoring.go]: Lines 778-779 duplicate 788-790. Also missing "This field is optional." present on all siblings. Please align with the pattern in e.g. `OpenShiftStateMetricsConfig.Resources`:

```suggestion
	// This field is optional.
	// More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
	// This is a simplified API that maps to Kubernetes ResourceRequirements.
```

File: config/v1alpha1/types_cluster_monitoring.go (hunk 20 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go:795]: Good catch, thanks! Added the three-test ratcheting pattern (retain via ratcheting, reject changed invalid list, allow reducing to valid count) for prometheusConfig.resources with initialCRDPatches, matches the existing coverage for the other component configs.

File: config/v1alpha1/types_cluster_monitoring.go (hunk 22 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go]: Thanks for catching that. Removed the duplicate lines and aligned with the OpenShiftStateMetricsConfig.Resources pattern — added This field is optional., More info: link, and This is a simplified API... line.

File: config/v1alpha1/types_cluster_monitoring.go (hunk 21 above)
Comment (danielmellado) [reply config/v1alpha1/types_cluster_monitoring.go]: Done!

**Review by saschagrunert — APPROVED** [review]
/lgtm

File: config/v1alpha1/types_cluster_monitoring.go [hunk 23]
```diff
@@ -202,13 +208,13 @@ type AlertmanagerCustomConfig struct {
...
-	// Maximum length for this list is 10.
+	// Maximum length for this list is 5.
 	// Minimum length for this list is 1.
 	// Each resource name must be unique within this list.
 	// +optional
 	// +listType=map
 	// +listMapKey=name
-	// +kubebuilder:validation:MaxItems=10
+	// +kubebuilder:validation:MaxItems=5
 	// +kubebuilder:validation:MinItems=1
```
[the remark this answers is not in the document]
Comment (everettraven) [reply config/v1alpha1/types_cluster_monitoring.go:305]: v1alpha1 API which never has to handle upgrade scenarios because we never ship v1alpha1 APIs in the Default feature set, only *NoUpgrade feature sets.

**Review by everettraven — APPROVED** [review]
/lgtm

**Comment by everettraven** [comment]
Alpha API with a reasonable reason for changing from 10 -> 5 max items on container resources fields.

/override ci/prow/verify-crdify
